keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt

```diff
@@ -71,6 +78,45 @@ class ResourceActuator(
 
         val response = vetoEnforcer.canCheck(resource)
         if (!response.allowed) {
+          /**
+           * [ArtifactVersioned] is a special [resource] sub-type. When a veto response sets
+           * [VetoResponse.vetoArtifact] and the resource under evaluation is of type
+           * [ArtifactVersioned], blacklist the desired artifact version from the environment
+           * containing [resource]. This ensures that the environment will be fully restored to
+           * a prior good-state.
+           */
+          if (response.vetoArtifact && resource.spec is ArtifactVersioned) {
+            try {
+              val (version, artifact) = when (desired) {
+                is Map<*, *> -> {
```

Comment:
Why the stars here if we're casting to a specific type below?

Reply:
It seemed required, please feel free to suggest a change.

Reply:
If you're specifically asking why it isn't `is Map<String, ArtifactVersioned>`, that doesn't work.